tests/first_stage_test.py

```diff
@@ -50,3 +336,128 @@ def test_valid_syntax(self):
+            stdin=testlib.subprocess.PIPE,
+        )
+
+        # Do not send all of the data from the preamble
+        proc.stdin.write(self.preamble[:-128])
+        proc.stdin.flush()
+        proc.stdin.close()
+        try:
+            returncode = proc.wait(timeout=10)
+        except testlib.subprocess.TimeoutExpired:
+            proc.kill()
+            proc.wait(timeout=3)
+            self.fail("First stage did not handle EOF on STDIN")
+        try:
+            self.assertEqual(0, returncode)
+            self.assertEqual(
+                proc.stdout.read(),
+                mitogen.parent.BootstrapProtocol.EC0_MARKER + b("\n"),
+            )
+            self.assertIn(
+                b("Error -5 while decompressing data"),
+                proc.stderr.read(),
+            )
+        finally:
+            proc.stdout.close()
+            proc.stderr.close()
+
+    def test_closed_stdin(self):
+        """This test closes STDIN of the child process.
+
+        1. The child process detects that STDIN is unavailable
+        2. The child process terminates early with an OSError exception, and
+           reports the issue via exception printed on STDERR.
+        3. The parent process correctly identifies this condition.
+
+        """
+        # We do not want to wait the default of 10s, change it to 0.1s
+        self.conn._first_stage_timeout = 0.1
+        args = self.conn.get_boot_command()
+
+        proc = testlib.subprocess.Popen(
+            args=args,
+            stdout=testlib.subprocess.PIPE,
+            stderr=testlib.subprocess.PIPE,
+            preexec_fn=lambda: os.close(0),
+            close_fds=True,
+        )
+        try:
+            stdout, stderr = proc.communicate(timeout=12)
+        except testlib.subprocess.TimeoutExpired:
+            proc.kill()
+            proc.wait(timeout=3)
+            self.fail("Closed STDIN situation was not recognized")
+        self.assertEqual(1, proc.returncode)
+        self.assertEqual(stdout, b"")
+        self.assertIn(
+            b("Bad file descriptor"),
+            stderr,
+        )
+
+    def test_closed_stdout(self):
+        """Test that first stage bails out if STDOUT is closed
+
+        This test closes STDOUT of the child process.
+
+        1. The child process detects that STDOUT is unavailable
+        2. The child process terminates early with an OSError exception, and
+           reports the issue via exception printed on STDERR.
+        3. The parent process correctly identifies this condition.
+
+        """
+
+        stdout_r, stdout_w = mitogen.core.pipe()
+        mitogen.core.set_cloexec(stdout_r.fileno())
+        stderr_r, stderr_w = mitogen.core.pipe()
+        mitogen.core.set_cloexec(stderr_r.fileno())
+        try:
+            proc = testlib.subprocess.Popen(
+                args=self.args,
+                stdout=stdout_w,
+                stderr=stderr_w,
+                preexec_fn=lambda: os.close(0),
```

Comment:
Possible copy paste error. Did you mean `os.close(1)`?

Reply:
Yesp will use os.close(pty.STDOUT_FILENO) in the new PR.